### PR TITLE
Ensure that buffer index is not -1 in allocation_planner

### PR DESCRIPTION
**Pull request by pranav-prakash**
See issue #7020

### CODE REVIEW

File: onnxruntime/core/framework/allocation_planner.cc [hunk 1]
```diff
@@ -299,6 +299,7 @@ class PlannerImpl {
           if (p_input_arg->Exists()) {
             auto input_arg_index = Index(p_input_arg->Name());
             auto original = Buffer(input_arg_index);
+            ORT_ENFORCE(original != -1, "Could not find buffer corresponding to input: " + p_input_arg->Name());
             if (1 == UseCount(original)) {
```
Comment (skottmckay) [inline onnxruntime/core/framework/allocation_planner.cc:303]: Why is this necessary? If there's a specific problem a unit test should be added to replicate the issue.

Comment (pranav-prakash) [reply onnxruntime/core/framework/allocation_planner.cc:303]: See the linked issue. Certain invalid models can cause the allocation planner to throw an exception
```
n >= 0 && static_cast<size_t>(n) < ort_value_info_.size() was false. 
```

I believe adding a more informative message is a better option.

Comment (pranav-prakash) [reply onnxruntime/core/framework/allocation_planner.cc:303]: Resolving for now

Comment (skottmckay) [reply onnxruntime/core/framework/allocation_planner.cc:303]: I would expect model loading is a better place to do such checks. 

If I understand correctly, the buffer lookup fails due to ProcessDefs not being called on the value, which can only happen if the value is unknown (not a graph input, initializer, or output of a previous node). 

I tried to repro with the model in #7020 but that fails due to a different error (the model uses ONNX opset 10, but MaxPool did not support quantized types in that opset version: https://github.com/onnx/onnx/blob/master/docs/Changelog.md#MaxPool-10)

>    [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Type Error: Type 'tensor(uint8)' of input parameter (7_quantized) of operator (MaxPool) in node (8) is invalid.

If we need to fix this, most likely it should be around here and not in the allocation planner: https://github.com/microsoft/onnxruntime/blob/368e4a324f5235823cd8a948c505536b13e31ada/onnxruntime/core/graph/graph.cc#L1484-L1485

Comment (pranav-prakash) [reply onnxruntime/core/framework/allocation_planner.cc:303]: @skottmckay The attached model does not seem to have a MaxPool with an input of `7_quantized`. The model graph has `DequantizeLinear -> Maxpool`

```
input: "7_scale"
input: "7_zero_point"
output: "8"
name: "8_DequantizeLinear"
op_type: "DequantizeLinear"

 input: "8"
output: "9"
name: "8"
op_type: "MaxPool"
```

I tested with onnxruntime 1.7.0 installed from pip and was able to repro the issue. Is it possible that one of the more recent commits to OR added QDQ fusion for maxpool?

----

Comment (skottmckay) [reply onnxruntime/core/framework/allocation_planner.cc:303]: Must have. Was able to load the model with optimizations turned off.

I created a PR to fix this during model loading. #7459

Comment (pranav-prakash) [reply onnxruntime/core/framework/allocation_planner.cc:303]: @skottmckay Great thanks! I created a separate issue for the QDQ opset issue you observed (#7460). I'll go ahead and close this issue since your PR supersedes this.
